src/caches/beliefpropagationcache.jl

```diff
@@ -0,0 +1,221 @@
+default_mt_constructor(inds_e) = ITensor[denseblocks(delta(inds_e))]
+default_mt_storage(ptn::PartitionedGraph) = Dict()
+function default_contractor(contract_list::Vector{ITensor}; kwargs...)
+  return contract_exact(contract_list; kwargs...)
+end
+default_contractor_kwargs() = (; normalize=true, contraction_sequence_alg="optimal")
+@traitfn default_bp_niters(g::::(!IsDirected)) = is_tree(g) ? 1 : nothing
+@traitfn function default_bp_niters(g::::IsDirected)
+  return default_bp_niters(undirected_graph(underlying_graph(g)))
+end
+function message_tensor_diff(mt_a::Vector{ITensor}, mt_b::Vector{ITensor})
+  LHS, RHS = contract(mt_a), contract(mt_b)
+  return 0.5 *
+         norm((denseblocks(LHS) / sum(diag(LHS))) - (denseblocks(RHS) / sum(diag(RHS))))
+end
+
+struct BeliefPropagationCache{PTN,MTS}
+  partitioned_itensornetwork::PTN
+  message_tensors::MTS
+end
+
+#Constructors...
+function BeliefPropagationCache(ptn::PartitionedGraph; mt_storage=default_mt_storage)
+  mts = default_mt_storage(ptn)
+  return BeliefPropagationCache(ptn, mts)
+end
+
+function BeliefPropagationCache(tn::ITensorNetwork, partitioned_vertices; kwargs...)
+  ptn = PartitionedGraph(tn, partitioned_vertices)
+  return BeliefPropagationCache(ptn; kwargs...)
+end
+
+partitioned_itensornetwork(bpc::BeliefPropagationCache) = bpc.partitioned_itensornetwork
+message_tensors(bpc::BeliefPropagationCache) = bpc.message_tensors
+function tensornetwork(bpc::BeliefPropagationCache)
+  return unpartitioned_graph(partitioned_itensornetwork(bpc))
+end
+function NamedGraphs.partitioned_graph(bpc::BeliefPropagationCache)
+  return partitioned_graph(partitioned_itensornetwork(bpc))
+end
+
+function initial_message_tensor(
+  bpc::BeliefPropagationCache, edge::PartitionEdge; mt_constructor=default_mt_constructor
+)
+  ptn = partitioned_itensornetwork(bpc)
+  src_e_itn = subgraph(ptn, src(edge))
+  dst_e_itn = subgraph(ptn, dst(edge))
+  inds_e = commoninds(src_e_itn, dst_e_itn)
+  return mt_constructor(inds_e)
+end
```

Comment:
I think we could just call this `default_message_tensor`.

Maybe we could also shorten this function by defining a function:
```julia
function linkinds(bpc::BeliefPropagationCache, edge::PartitionEdge)
  ptn = partitioned_itensornetwork(bpc)
  src_e_itn = subgraph(ptn, src(edge))
  dst_e_itn = subgraph(ptn, dst(edge))
  return commoninds(src_e_itn, dst_e_itn)
end

function default_message_tensor(
  bpc::BeliefPropagationCache, edge::PartitionEdge
)
  return default_message_tensor(bpc)(linkinds(bpc, edge))
end
```
also making use of my suggestion above to store the `default_message_tensor` constructor inside the `BeliefPropagationCache`.

Reply:
True although there isn't actually a `partitionedITensorNetwork` type?

Reply:
Could be defined for `PartitionedGraph`.

Reply:
But isn't that strange to have a function `f(a)` which relies on some field in `a` being of a certain type (which the actual struct `a` is not restricted to)?

Reply:
I think that is more a philosophical question about coding style and generic code design, but in practice I don't see a problem with doing that in this case. The function will just fail farther down if the partitioned graph isn't wrapping an underlying graph which is an `AbstractITensorNetwork`, or something that has the interface of an `AbstractITensorNetwork`. In general I have been leaning towards being less restrictive about typing, and focusing more on function interfaces (i.e. which functions work on an object, rather than which type it is or what it is a subtype of, which is called [duck typing](https://en.wikipedia.org/wiki/Duck_typing)).

Reply:
We can always add some more restrictions later if needed, for example by defining a `PartitionedITensorNetwork` type or making a trait, but I think that's only needed if there is some ambiguity that arises, for example if there is some other reasonable definition of `linkinds` for a `PartitionedGraph` that is wrapping something else besides an `AbstractITensorNetwork`, which require a specialized implementation not covered by the current implementation.

In general, I've found that it is better to start out being as unrestrictive on type constraints of functions (and fields of types) as possible, which forces you to think about interfaces rather than focusing too much on the particular objects you have in mind, which ultimately makes the code more generic and concise in the long run.

Reply:
If anything, I would lean towards making the definition even more generic, and define it as:
```julia
function linkinds(tn, edge::PartitionEdge)
  tn_src = subgraph(tn, src(edge))
  tn_dst = subgraph(tn, dst(edge))
  return commoninds(tn_src, tn_dst)
end
```
Then, this function works for any type that defines `subgraph(tn, vertex::PartionVertex)`, which could be `BeliefPropagationCache`, `AbstractITensorNetwork`, or something else.

Reply:
Yeah, that sounds like a good idea (though better to call it `src/partitioneditensornetwork.jl`, i.e. make it lower case).